enhancements/olm/safe-cluster-upgrades.md

```diff
@@ -0,0 +1,270 @@
+                "version": "v0.0.2",
+                "channels": ["alpha"],
+                "maxOpenShiftVersion": "4.7.0", # Prevent upgrades to 4.8
+            }
+        ],
+        "defaultChannel": "alpha",
+        "icon": "asdfasdfoipuasdfvasdfqer",
+        "channelUpgradeGraphs": {
+            "channels": [
+                {
+                    "name": "alpha",
+                    "upgrades": [
+                        {
+                            "to": {
+                                "bundle": "quay.io/random/etcdv0.0.1"
+                            }
+                        },
+                        {
+                            "from": {
+                                "bundles": ["quay.io/random/etcdv0.0.1"]
+                            },
+                            "to": {
+                                "bundle": "quay.io/somethingelse/etcdv0.0.2"
+                            }
+                        }
+                    ]
+                }
+            ]
+        }
+    }
+```
+
+Like all bundle properties, these properties would be propagated to the CSV as an annotation. By specifying the MaxOpenShiftVersion using the Declarative Index Config there will be no need to publish a new version of the operator to modify the maxOpenShiftVersion annotation once it is tested against new OpenShift Versions. Instead the bundle defined in the Declarative Index Config would just need to have the maxOpenShiftVersion set to the desired value for each bundle.
+
+#### How OLM will determine if it should block upgrades
+
+When OLM is reconciling CSV it will check the set of annotations for an annotation whose key matches `operators.coreos.com/maxOpenShiftVersion` and the value is the maximum version of OpenShift that the operators supports. An example of this annotation can be seen below:
+
+```yaml=
+apiVersion: operators.coreos.com/v1alpha1
+kind: ClusterServiceVersion
+metadata:
+  annotations:
+    # Prevent upgrades to OpenShift Version 4.9
+    operators.coreos.com/maxOpenShiftVersion: 4.8.0 
+```
+
+> NOTE: OLM should allow users to specify Major.Minor.Patch versions in this field but OLM can only prevent Major.Minor OpenShift upgrades. By allowing Major.Minor.Patch versions to be specified, OLM will not need to modify the existing annotations if OpenShift ever allows ClusterOperators to prevent patch upgrades.
+
+Assume that the CSV shown above is the only CSV on the Cluster. In this case, OLM would compare the value of the `operators.coreos.com/maxOpenShiftVersion` annotation against the version of the Cluster to decide it's upgradeable status.
+
+Examples can be found below:
+
+| OpenShift Cluster Version   | CSV's Max OpenShift Version | Is Upgradeable?|
+| --------------------------- | --------------------------- | -------------- |
+| 4.6.z                       | 4.8.0                       | True           |
+| 4.7.z                       | 4.8.0                       | True           |
```

Comment:
This enhancement proposal claims to be asking for patch-level `operators.coreos.com/maxOpenShiftVersion` values above ("By allowing Major.Minor.Patch versions to be specified...").  But if you're running 4.7.30 (or whatever), your only 4.8 update targets might be 4.8.4 (or whatever).  That would be over the claimed 4.8.0 max.  If you want to keep the door open for patch-level specificity (because you don't trust the OpenShift core to preserve [the required backwards compatibility][1]?), you should probably give folks a way to say "I'm ok floating without claiming a particular patch will break me" like `4.8.*`.  If you're ok having everyone float, you can restrict the `maxOpenShiftVersion` to just `{major}.{minor}`.

[1]: https://semver.org/spec/v2.0.0.html#spec-item-6

Reply:
This is a good point, @wking.  Given we can't (won't) actually enforce the z-stream level, we should probably not allow it to be specified in the field.

if that changes someday, we can loosen the validation on the field.

Reply:
Yeah, above it's stated that this is meant to allow a future state where we'd allow inhibition of patch upgrades without changing existing annotations but that's only true if they chose an inflated value from the start, ie: 4.8.999999

While there have been past proposals for `UpgradeableMicro` or `UpgradeablePatch` to inhibit patch version upgrades those have been rejected because we always want to allow ourselves the ability to alter the logic for setting `Upgradeable` in a later patch version if necessary. If we ever inhibit patch versions we then pushing admins toward forcing an upgrade and we never want to encourage that.

Reply:
Understood and thank you for the information. I will update the PR to only allow `{major}.{minor}` declarations.

Reply:
Addressed in the latest version of the enhancement.

Reply:
Please mark as resolved if applicable.